### PR TITLE
Catch errors while coercing 'and' intro patterns

### LINKED ISSUE #6384: Anomaly: Not_found from using an identifier from intros
Bug originally reported by [Bruno Le Floch on coq-club](https://sympa.inria.fr/sympa/arc/coq-club/2017-12/msg00045.html).

#### Version

a77f3a3e0

The Coq Proof Assistant, version 8.8+alpha (December 2017)
compiled on Dec 11 2017 9:49:14 with OCaml 4.04.2

#### Operating system

macOS

#### Description of the problem

The following results in an anomaly. Note that `as H` is an invalid intro pattern, but happens to overlap with the variable introduced by `intros H`.

```coq
Theorem test (A:Prop) : A \/ A -> A.
  Fail intro H; destruct H as H.
  (* Error: Disjunctive/conjunctive introduction pattern expected. *)
  intros H; destruct H as H.
Abort.
```

### LINKED ISSUE #6385: Uncaught exception CannotCoerceTo when using a tactic as an intro pattern
#### Version

a77f3a3e0

The Coq Proof Assistant, version 8.8+alpha (December 2017)
compiled on Dec 11 2017 9:49:14 with OCaml 4.04.2

#### Operating system

macOS

#### Description of the problem

The following results in `Anomaly
"Uncaught exception Ltac_plugin.Taccoerce.CannotCoerceTo("an introduction pattern")."`. It's raised by `taccoerce.ml`'s `coerce_to_intro_pattern` and uncaught by `tacinterp.ml`'s `interp_or_and_intro_pattern_option`.

```coq
Theorem test (A:Prop) : A \/ A -> A.
  Fail let H := idtac in intros H; destruct H as H'.
  (* Disjunctive/conjunctive introduction pattern expected. *)
  let H' := idtac in intros H; destruct H as H'.
Abort.
```

### PR DESCRIPTION
**Pull request by tchajed**
Fixes #6384.
Fixes #6385.

### CODE REVIEW

File: plugins/ltac/tacinterp.ml [hunk 1]
```diff
@@ -890,10 +890,9 @@ let interp_intro_pattern_naming_option ist env sigma = function
 let interp_or_and_intro_pattern_option ist env sigma = function
   | None -> sigma, None
   | Some (ArgVar (loc,id)) ->
-      (match coerce_to_intro_pattern env sigma (Id.Map.find id ist.lfun) with
-      | IntroAction (IntroOrAndPattern l) -> sigma, Some (loc,l)
-      | _ ->
-        user_err ?loc (str "Cannot coerce to a disjunctive/conjunctive pattern."))
+    (match interp_intro_pattern_var loc ist env sigma id with
```
Comment (psteckler) [inline plugins/ltac/tacinterp.ml]: This could be formatted so the only actual difference, the match discriminee, gets marked by Github.

Comment (tchajed) [reply plugins/ltac/tacinterp.ml]: Good point, done (this also caught that I accidentally dropped a period in the error message).

**Comment by Zimmi48** [comment]
@herbelin Do you get a clear e-mail notification when someone ask for your review? If yes, sorry for the noise. If no, I just want to make sure that you see this request for review because PRs that are not approved today won't make it into 8.7.1.

**Review by ppedrot — APPROVED** [review]
In case @herbelin doesn't make it on time: this is crazy Ltac semantics, but the fix is correct relatively to that semantics.

**Comment by herbelin** [comment]
I got the mail asking a review but I did not see it before now. Good that @ppedrot could do it.

File: test-suite/output/InvalidDisjunctiveIntro.out [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+The command has indeed failed with message:
+Cannot coerce to a disjunctive/conjunctive pattern.
+The command has indeed failed with message:
+Ltac variable H is bound to <tactic closure> which cannot be coerced to
+an introduction pattern.
+The command has indeed failed with message:
+Disjunctive/conjunctive introduction pattern expected.
+The command has indeed failed with message:
+Ltac variable H' is bound to <tactic closure> which cannot be coerced to
+an introduction pattern.
```
Comment (Zimmi48) [inline test-suite/output/InvalidDisjunctiveIntro.out]: FTR the output in 8.7 is:

```
The command has indeed failed with message:
Ltac variable H' is bound to a tactic which cannot be
coerced to an introduction pattern.
```

I think this change in the error message is linked to 565a9a1b5368c586e529fc9774e4cb4b81c6441b. I'm just writing this comment to ensure that this change of `a tactic` to `<tactic closure>` in the error message was intended (cc @herbelin).

Comment (herbelin) [reply test-suite/output/InvalidDisjunctiveIntro.out]: Well, it is a generic printer and I did not know what to write. For instance, the message is  used also for `idtac x y z` where we would probably not write "a tactic" if some of `x`, `y`, or `z` is bound to a tactic.

I don't know what to to. Maybe going back to an ad hoc printer for this particular case?

Comment (tchajed) [reply test-suite/output/InvalidDisjunctiveIntro.out]: I think it's fine as-is with "<tactic closure>" - consistency between error messages is nice, both from a user and code base perspective.

Comment (Zimmi48) [reply test-suite/output/InvalidDisjunctiveIntro.out]: Yeah, I don't think the message is problematic either. I was just making sure that the change did not happen by accident.

Comment (herbelin) [reply test-suite/output/InvalidDisjunctiveIntro.out]: @tchajed: So writing `"a tactic closure"` (with the quotes) would be ok for you?

Comment (tchajed) [reply test-suite/output/InvalidDisjunctiveIntro.out]: I actually like the current error message, with the angle brackets. It emphasizes that there's a (currently unprintable) tactic closure assigned to that variable.

Comment (herbelin) [reply test-suite/output/InvalidDisjunctiveIntro.out]: Well, so let's keep it as it is, at least for the moment.

**Pull request merged**